### ISSUE TITLE
Proposal: "first" option for `indent/MemberExpression`

### ISSUE THREAD

**Issue by ojab** [eslint/eslint#9383]
**Tell us about your environment**

* **ESLint Version:** 4.8.0
* **Node Version:** 8.6.0
* **npm Version:** yarn-1.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What rule do you want to change?**
indent

**Does this change cause the rule to produce more or fewer warnings?**
nope

**How will the change be implemented? (New option, new default behavior, etc.)?**
Option `{ "MemberExpression": "first" }` will allow member expressions to be indented just as the first member expression.

**Please provide some example code that this change will affect:**

```js
foo.bar
   .baz
   .qux
```

**What does the rule currently do for this code?**
Depends on configuration, but `.baz` & `.qux` are indented by fixed number of spaces.

**What will the rule do after it's changed?**
`.baz` & `.qux` will be indented as shown, i. e. dots before member name will be on the same column.

**Comment by platinumazure** [comment]
Sounds good to me. I'll champion.

@eslint/eslint-team Any :+1:s for this proposal?

**Comment by akouryy** [comment]
I want this option. Are there any updates on it? Waiting for evaluation?

**Comment by platinumazure** [comment]
@eslint/eslint-team Anyone want to :+1: this proposal? This should be a pretty easy change since we already support "first" indentation for other lists (object properties, array items, function parameters).

Although: Should we actually align with the last one on a line?

```js
this.someObject.someMethod()
                          .someOtherMethod();
```

@ojab Thoughts on the above?

**Comment by not-an-aardvark** [comment]
> This should be a pretty easy change since we already support "first" indentation for other lists (object properties, array items, function parameters).

I think the change would be nontrivial because `MemberExpression` indentation is not handled in the same way as properties, array items, and function parameters. However, it does seem like it should be possible without a huge amount of difficulty.

Additional case to consider:

```js
a.b(
    x
)
 .c
```

**Comment by AlexandreBonneau** [comment]
In my opinion, if such option is set and the indent is 4 spaces, then (where * is a space)
```js
foo.b(
****x
)
***.c
```
should be aligned like so:
```js
foo.b(
*******x
***)
***.c
```
like some IDE do (ie. PhpStorm):
![js_chained_alignment](https://user-images.githubusercontent.com/7248877/34591832-4a2d385c-f164-11e7-81d2-e4a1d56e6413.png)

**Comment by platinumazure** [comment]
@ojab @AlexandreBonneau As we can see, there are a few possible interpretations for how "first" indentation could/should work. Can we discuss how it might look to either support both in the rule, or to definitely choose one interpretation or the other based on their use in common style guides? Thanks!

**Comment by ojab** [comment]
I'm intereseted in methods that are aligned to the first call, i. e.:
```javascript
this.someObject.someMethod()
    .someOtherMethod()
```
and I agree with @AlexandreBonneau here about indent of nested code:
```javascript
foo.b(
*******x
***)
***.c
```

**Comment by ljharb** [comment]
@ojab wouldn't the first call mean it's like this?:
```js
this.someObject.someMethod()
               .someOtherMethod()
```

**Comment by ojab** [comment]
Nope, I meant not the first method call, but first [I don't really know how it's called in JS/eslint, MemberExpression?] anything call, i. e. first dot.

**Comment by ljharb** [comment]
Hm, I would assume that some people might want "first property access" and some might want "first called method".

**Comment by ojab** [comment]
Yeah, it can be configuration option.

**Comment by not-an-aardvark** [comment]
I feel like the desired behavior here isn't well-defined in general -- property accesses and calls are not the only operations that can be taken on `MemberExpression` nodes.

A few more cases:

```js
foo.bar + baz.qux
  .boop
```

```js
(foo.bar + baz.qux)
  .boop
```

```js
(function() { foo.bar }).baz.qux
  .boop
```

**Comment by AlexandreBonneau** [comment]
About the simplest cases:
```js
// 1)
this.someObject.someMethod()
               .someOtherMethod();
someObject.someMethod()
          .someOtherMethod();
```
or
```js
// 2)
this.someObject.someMethod()
    .someOtherMethod()
```
it seems to me that `1)` is the most used case, and that if the alignment should be on the very first dot, then `2)` could/should be reformatted to:

```js
// 2b)
this.someObject
    .someMethod()
    .someOtherMethod();
```

**Comment by AlexandreBonneau** [comment]
Is there anything we can do to make this happen, if accepted as an enhancement?

**Comment by ljharb** [comment]
ftr, the airbnb style guide would love to use this option, provided it could be configured to *not* align with anything except normal semantic indentation levels. In particular, we want things like:
```js
Promise.resolve()
  .then()
  .catch();

Object.keys()
  .filter()
  .map();
```
and we explicitly do *not* want this:
```js
Promise
  .resolve()
  .then()
  .catch();

Object
  .keys()
  .filter()
  .map();
```
It’s notable that Prettier outputs the latter block, so having an autofixable rule to correct this would be wonderful.

**Comment by not-an-aardvark** [comment]
@ljharb I'm confused, isn't that the current behavior of the `indent` rule as far as indentation is concerned? It seems like you want a new rule disallowing a linebreak before the first property in a "chain", but that doesn't seem related to this issue.

**Comment by ljharb** [comment]
@not-an-aardvark ah, perhaps I'm getting confused about which rule does what :-)

**Comment by joshua-golub** [comment]
Regarding "A few more cases" from @not-an-aardvark:

    foo.bar + baz.qux
      .boop
    
    ---> becomes <---

    foo.bar + baz.qux
                 .boop
and

    (foo.bar + baz.qux)
      .boop
    
    ---> is unchanged because there are no property chains <---

and

    (function() { foo.bar }).baz.qux
      .boop
    
    ---> becomes <---
    
    (function() { foo.bar }).baz.qux
                            .boop

**Comment by joshua-golub** [comment]
@AlexandreBonneau, do you mean prevents it from getting formatted like that with the **current** implementation of `indent`?

If so, then yes. Meaning: I can't find any configuration of `indent` that allows this:

    (function() { foo.bar }).baz
                            .qux
                            .boop 

or, for that matter:

    foo.bar
       .baz
       .qux
       .boop

**Comment by platinumazure** [comment]
Removing my assignment/championship until we can get more clarity around how this should behave.

**Comment by AlexandreBonneau** [comment]
Ok let's recap to see how that indent modification proposal would work :wink:

### The proposed solution

The way I see it, there are two places you usually align your members; either the very *first* or very *last* dot of the **first line**.
Also, some people like to chain multiple members *on the same line*, while others want each member on its on line, and others just want to put newlines everywhere even before the first member.

That means we could separate those 2 choices, and that will create 4 different ways of aligning the code (plus a 5th way where each member is put on a new line).

*Note: For all the subsequent examples, I'll use a 4 spaces indentation*

```js
// 1) ~~~~~~~ First dot, same line
Promise.resolve().then()
.catch().foo().bar();

// will return:
Promise.resolve().then()
       .catch().foo().bar();

// 2) ~~~~~~~ First dot, different lines
Promise.resolve().then()
.catch().foo().bar();

// will return:
Promise.resolve()
       .then()
       .catch()
       .foo()
       .bar();

// 3) ~~~~~~~ Last dot, same line
Promise.resolve().then().catch()
.foo().bar();

// will return:
Promise.resolve().then().catch()
                        .foo().bar();

// 4) ~~~~~~~ Last dot, different lines
Promise.resolve().then().catch()
.foo().bar();

// will return:
Promise.resolve().then().catch()
                        .foo()
                        .bar();

// 5) ~~~~~~~ Newlines everywhere
Promise.resolve().then()
.catch().foo().bar();

// will return:
Promise
    .resolve()
    .then()
    .catch()
    .foo()
    .bar();
```

This *should* cover every cases.

With the example from @joshua-golub:
```js
(function() { foo.bar }).baz.qux
  .boop

// Would return:
// 1) First dot, same line
(function() { foo.bar }).baz.qux
                        .boop;
// 2) First dot, different lines
(function() { foo.bar }).baz
                        .qux
                        .boop;
// 3) Last dot, same line
(function() { foo.bar }).baz.qux
                            .boop;
// 4) Last dot, different lines
(function() { foo.bar }).baz.qux
                            .boop;
// 5) Newlines everywhere
(function() { foo.bar })
    .baz
    .qux
    .boop;
```

Do note that I think in all the cases except number 5, *the first member should put on the very first line*, even if there is a newline initially before it:
```js
(function() { foo.bar })
 .baz.qux
  .boop

// Would return exactly the same thing than above:
// 1) First dot, same line
(function() { foo.bar }).baz.qux
                        .boop;
// 2) First dot, different lines
(function() { foo.bar }).baz
                        .qux
                        .boop;
// 3) Last dot, same line
(function() { foo.bar }).baz.qux
                            .boop;
// 4) Last dot, different lines
(function() { foo.bar }).baz.qux
                            .boop;
// 5) Newlines everywhere
(function() { foo.bar })
    .baz
    .qux
    .boop;
```



Now, I'll let the eslint devs think of the best way to include those options.

I'll just write down what *could* be the structure of those *(please correct me there!)*:
```js
const options = {
    indent: {
        chaining: {
            dot: 'first', // or 'last'
            lines: 'same', // or 'newline' or 'always'
        }
    }
}
```
*`indent.chaining.dot` would be ignored when using `indent.chaining.lines === 'always'`*

This is the simplest way for me to make everybody happy with this new feature.

### The long-term improvement

You could want to give the users more control over there indentation, for instance by adding a variant where the alignment would be done on the first dot following a particular member type (ie. `object`, `function` or `attribute`).
I'm guessing that would complexify that feature quite a bit, but I could be wrong.
This is why I think this should be out of scope for now in order to speed things up on this feature.

Let's make it happen ;)

**Comment by AlexandreBonneau** [comment]
Not mixing indent with newlines sounds logical.

While on my end I would always use the case `2)` I've mentioned above (*First dot, different lines*), how would you with those simplified 3 cases mimic the case `3)` *First dot, same line*?

**Comment by nzakas** [comment]
@platinumazure you championed this proposal, is this something we should still be looking into?

**Comment by platinumazure** [comment]
> @platinumazure you championed this proposal, is this something we should still be looking into?

I think I had championed it under the assumption it would be pretty simple, but later discussion shows this is not the case. I'm not sure I want to add all of the variations that have been proposed here into the `indent` rule, which is already very complicated as is.

I'm going to step down as champion. Not sure if this should be closed immediately or if we want to see if another team member wants to take this on.

**Comment by AlexandreBonneau** [comment]
Ouch, the close button has been promptly pressed.

Since there is a real demand from users on this indent variant, shouldn't the issue still be opened, until a good soul starts to write the RFC?
Closing it like that seems a little bit hasty, no?

**Comment by platinumazure** [comment]
My apologies, I should have noted our (separate) issue closure policy. We find that if an issue is open for 21 or more days without an ESLint champion, or even after 90 days with an accepted proposal, the issue tends not to be solved at all, or only rarely.

This closure does not mean we're rejecting the feature outright. It just means we know it won't be done soon and we're just managing the backlog.

If someone can come up with an RFC, we will definitely look at it.
